Yanovich Aleksey

AlexeyYan opened the pull request.

AlexeyYan committed:
Added base classes and base functionality.

AlexeiBuzuma commented on rss_reader/parser.py:
```diff
@@ -0,0 +1,45 @@
+from storage import Article
+
+
+class Parser:
```
Я бы уточнил название класса. `Parser` звучит очень абстрактно

AlexeyYan committed:
Added "--verbose" feature and exeptions catching

Added logging and implemented "--verbouse" feature.
Added specific exceptions and implemented their catching.
Some small fixes.

AlexeyYan committed:
Added additional check for links in <enclosure> tag.

Other changes:
- Rewrite error messages.
- Remove 'feed_date' field from feed structure.
- Added new date save format.

AlexeyYan committed:
Added tests for main and parser modules.  Added json chema.

AlexeyYan committed:
Version 1.0: Iteration 1 Copleted

AlexeyYan committed:
Fix mistakes in README.md

AlexeyYan committed:
Version 2.0: Iteration 2 completed.

AlexeyYan committed:
Fix mistakes in README.md

AlexeiBuzuma commented on setup.py:
```diff
@@ -0,0 +1,19 @@
+from setuptools import setup
+
+setup(
```
после установки в чистый контейнер python:latest 
```python
# rss_reader
Traceback (most recent call last):
  File "/usr/local/bin/rss_reader", line 5, in <module>
    from rss_reader import start_app
  File "/usr/local/lib/python3.8/site-packages/rss_reader/__init__.py", line 1, in <module>
    from rss_reader.main import main
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 1, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

AlexeiBuzuma commented on rss_reader/exceptions.py:
```diff
@@ -0,0 +1,17 @@
+import sys
```
```bash
# rss_reader https://timesofindia.indiatimes.com/rssfeeds/65857041.cms
Traceback (most recent call last):
  File "/usr/local/bin/rss_reader", line 8, in <module>
    sys.exit(start_app())
  File "/usr/local/lib/python3.8/site-packages/rss_reader/__init__.py", line 5, in start_app
    main()
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 72, in main
    reader.get_feed_from_source()
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 40, in get_feed_from_source
    self.feed = self.parser.parse(request.content)
  File "/usr/local/lib/python3.8/site-packages/rss_reader/feed_parser.py", line 37, in parse
    date = datetime.strptime(html.unescape(article.find('pubDate').text), "%a, %d %b %Y %H:%M:%S %z")
  File "/usr/local/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/usr/local/lib/python3.8/_strptime.py", line 349, in _strptime
    raise ValueError("time data %r does not match format %r" %
ValueError: time data 'Mon, 25 Nov 2019 03:21:05 GMT' does not match format '%a, %d %b %Y %H:%M:%S %z'
```

AlexeyYan committed:
Implemented functionality for Iteration 3.

Implemented caching of news while reading.
Other changes:
- Fixed bug with handling of date of RSS feed.
- Fixed bug with handling images without description.
- Fixed bug with installing third-party modules.

AlexeiBuzuma commented on rss_reader/exceptions.py:
```diff
@@ -0,0 +1,17 @@
+class SourceConnectingError(RSS_reader_error):
+    def __init__(self, source, msg):
+        print(f"SourceConnectingError: Can't get feed from '{source}': {msg}")
+        sys.exit(1)
```
Это очень неожиданный подход. 
В исключениях не нужно делать sys.exit 
у пользователей этого кода должна быть минимальная возможность попытаться обработать данное исклчение :) 
А в таком случае приложение будет сразу закрываться, когда было исключение (кроме ситуаций обработки SysExit)

AlexeyYan committed:
Iteration 4 completed

Added --to-fb2 option.
Added --to-html option.
Fix exceptions.

AlexeyYan committed:
Update README

AlexeiBuzuma commented:
Падает с эксепшеном при попытке сгенирировать файл
```
# rss_reader https://news.tut.by/rss/index.rss --to-fb2 .
Traceback (most recent call last):
  File "/usr/local/bin/rss_reader", line 8, in <module>
    sys.exit(start_app())
  File "/usr/local/lib/python3.8/site-packages/rss_reader/__init__.py", line 5, in start_app
    main()
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 131, in main
    reader.get_feed()
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 52, in get_feed
    self.coverter.convert_to_fb2(self.feed, self.fb2_file_path, self.limit)
  File "/usr/local/lib/python3.8/site-packages/rss_reader/converters.py", line 108, in convert_to_fb2
    binary.text = self._handle_image(image['source_url'])
  File "/usr/local/lib/python3.8/site-packages/rss_reader/converters.py", line 123, in _handle_image
    self.logger(f"Download image {image_url}")
TypeError: 'Logger' object is not callable
```

AlexeyYan committed:
Version 4.0 patch 1

- Fix all bugs with converting
- Fix bags with pathes validation
- Some changs in view of fb2

AlexeiBuzuma commented:
К сожалению, конвертация в fb2 падает 
```
# rss_reader "https://news.yahoo.com/rss/" --to-fb2 .
Traceback (most recent call last):
  File "/usr/local/bin/rss_reader", line 8, in <module>
    sys.exit(start_app())
  File "/usr/local/lib/python3.8/site-packages/rss_reader/__init__.py", line 5, in start_app
    main()
  File "/usr/local/lib/python3.8/site-packages/rss_reader/main.py", line 130, in main
    fb2_file_path = Path(cmd_args.to_html)
  File "/usr/local/lib/python3.8/pathlib.py", line 1018, in __new__
    self = cls._from_parts(args, init=False)
  File "/usr/local/lib/python3.8/pathlib.py", line 667, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/usr/local/lib/python3.8/pathlib.py", line 651, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

AlexeiBuzuma commented:
1) Не везде есть док стринги, хотя было бы уместно их добавить. 
2) Много магических чисел в коде (в освновном это индексы, но все же). Сделать код красивие в некоторых местах может помочть например namedtuple.
3) HTML шаблон создается обычной строкой в пайтон коде. Его можно вынести в отдельный файл. 
4) конвертация в fb2 находится в одной очень большой функции. Ее стоит декомпозировать. 


* хорошая идея с созданием XML файлов для тестов (чтобы было еще лучше, я бы предложил создать больше файлов под различные случаи)
